File: packages/ucm-core/src/markers/appendOnly.ts
// Append-only discipline check for JSONL ledgers (spec amendment 2; 4.3 rules
// 9/10). The core `appendOnly` is pure and line-based: the old content must be an
// unchanged PREFIX of the new content. Any edit, delete, or reorder of an
// existing line is a violation. The only impure piece is the thin git helper that
// reads the base-ref version of a file via `git show <base-ref>:<path>`.
import { execFileSync } from "node:child_process";

export interface AppendOnlyViolation {
  // "edited": an existing line's text changed (a reorder shows up as an edit).
  // "deleted": an existing line is gone because the new content is shorter.
  kind: "edited" | "deleted";
  index: number; // 0-based index into oldLines where the prefix first breaks
  old_line: string;
  new_line: string | null; // null when the line was deleted (new is shorter)
  message: string;
}

export type AppendOnlyResult = { ok: true } | { ok: false; violation: AppendOnlyViolation };

// Pure append-only check: oldLines must equal the leading prefix of newLines.
export function appendOnly(
  oldLines: ReadonlyArray<string>,
  newLines: ReadonlyArray<string>
): AppendOnlyResult {
  for (let i = 0; i < oldLines.length; i += 1) {
    if (i >= newLines.length) {
      return {
        ok: false,
        violation: {
          kind: "deleted",
          index: i,
          old_line: oldLines[i],
          new_line: null,
          message: `existing line ${i + 1} was deleted; the registry is append-only`
        }
      };
    }
    if (newLines[i] !== oldLines[i]) {
      return {
        ok: false,
        violation: {
          kind: "edited",
          index: i,
          old_line: oldLines[i],
          new_line: newLines[i],
          message: `existing line ${i + 1} was edited or reordered; the registry is append-only`
        }
      };
    }
  }
  return { ok: true };
}

// Split JSONL text into content lines, dropping a single trailing newline's empty
// segment. Interior blank lines are preserved so a blank-line edit is not hidden.
export function splitJsonlLines(text: string): string[] {
  if (text === "") {
    return [];
  }
  const lines = text.split("\n");
  if (lines.length > 0 && lines[lines.length - 1] === "") {
    lines.pop();
  }
  return lines;
}

// Injectable git runner for testing the base-ref read without a real repo.
export type GitRunner = (args: string[], cwd?: string) => string;

export interface ReadBaseRefOptions {
  cwd?: string;
  runner?: GitRunner;
}

const defaultRunner: GitRunner = (args, cwd) =>
  execFileSync("git", args, { cwd, encoding: "utf8" });

// Patterns git emits when a path does not exist at the given ref. Treated as
// "old content is empty" (a brand-new file added on this branch).
const ABSENT_AT_BASE = /does not exist in|exists on disk, but not in/i;

// Read the base-ref version of a file (`git show <base-ref>:<path>`). Returns ""
// when the path does not exist at the base ref (a newly added file). Any other
// git failure is rethrown with context. Thin and impure by design.
export function readBaseRefFile(
  baseRef: string,
  path: string,
  options: ReadBaseRefOptions = {}
): string {
  const runner = options.runner ?? defaultRunner;
  try {
    return runner(["show", `${baseRef}:${path}`], options.cwd);
  } catch (error) {
    const stderr = String(
      (error as { stderr?: unknown }).stderr ?? (error as Error).message ?? ""
    );
    if (ABSENT_AT_BASE.test(stderr)) {
      return "";
    }
    throw new Error(`git show ${baseRef}:${path} failed: ${stderr.trim() || "unknown error"}`);
  }
}

// Convenience: check current JSONL text against its base-ref version, splitting
// both into lines first.
export function appendOnlyAgainstBaseRef(
  baseRef: string,
  path: string,
  currentText: string,
  options: ReadBaseRefOptions = {}
): AppendOnlyResult {
  const oldText = readBaseRefFile(baseRef, path, options);
  return appendOnly(splitJsonlLines(oldText), splitJsonlLines(currentText));
}

File: UseCasesCore/Sources/UseCasesCore/Markers/AppendOnly.swift
import Foundation

public enum AppendOnlyViolationKind: String, Equatable, Sendable {
  /// An existing line's text changed; a reorder shows up as an edit.
  case edited
  /// An existing line is gone because the new content is shorter.
  case deleted
}

public struct AppendOnlyViolation: Equatable, Sendable {
  public let kind: AppendOnlyViolationKind
  /// 0-based index into the old lines where the prefix first breaks.
  public let index: Int
  public let oldLine: String
  /// Nil when the line was deleted.
  public let newLine: String?
  public let message: String

  var jsonValue: JSONValue {
    .object(JSONObject([
      ("kind", .string(kind.rawValue)),
      ("index", .number(Double(index))),
      ("old_line", .string(oldLine)),
      ("new_line", newLine.map(JSONValue.string) ?? .null),
      ("message", .string(message)),
    ]))
  }
}

public enum AppendOnlyResult: Equatable, Sendable {
  case holds
  case violated(AppendOnlyViolation)

  /// `{ ok: true }` or `{ ok: false, violation }`.
  var jsonValue: JSONValue {
    switch self {
    case .holds:
      .object(JSONObject([("ok", .bool(true))]))
    case let .violated(violation):
      .object(JSONObject([("ok", .bool(false)), ("violation", violation.jsonValue)]))
    }
  }
}

/// Append-only discipline for JSONL ledgers (appendOnly.ts): the old content
/// must be an unchanged PREFIX of the new. The only impure piece is the read of
/// a file's base-ref version through `git show`.
public enum AppendOnly {
  /// `oldLines` must equal the leading prefix of `newLines`, code unit for code
  /// unit — a canonically equivalent respelling is an edit.
  public static func check(
    oldLines: [String],
    newLines: [String],
  ) -> AppendOnlyResult {
    for (index, oldLine) in oldLines.enumerated() {
      guard index < newLines.count else {
        return .violated(AppendOnlyViolation(
          kind: .deleted,
          index: index,
          oldLine: oldLine,
          newLine: nil,
          message: "existing line \(index + 1) was deleted; the registry is append-only",
        ))
      }
      guard JavaScriptString.identical(newLines[index], oldLine) else {
        return .violated(AppendOnlyViolation(
          kind: .edited,
          index: index,
          oldLine: oldLine,
          newLine: newLines[index],
          message: "existing line \(index + 1) was edited or reordered; "
            + "the registry is append-only",
        ))
      }
    }
    return .holds
  }

  /// Content lines, dropping the single empty segment a trailing newline
  /// leaves. Interior blank lines are kept so a blank-line edit is not hidden.
  public static func splitJSONLines(_ text: String) -> [String] {
    guard !text.isEmpty else {
      return []
    }
    var lines = JavaScriptString.split(text, on: CodeUnits.lineFeed)
    if lines.last == "" {
      lines.removeLast()
    }
    return lines
  }

  /// `git show <base-ref>:<path>`; "" when the path does not exist at the ref
  /// (a file added on this branch). Any other git failure is raised.
  public static func readBaseReferenceFile(
    baseReference: String,
    path: String,
    workingDirectory: String?,
    runner: some GitRunning = GitProcessRunner(),
  ) throws(GitError) -> String {
    do throws(GitError) {
      return try runner.run(
        ["show", "\(baseReference):\(path)"],
        workingDirectory: workingDirectory,
      )
    } catch {
      let detail = error.standardError ?? error.message
      if isAbsentAtBase(detail) {
        return ""
      }
      let trimmed = JavaScriptString.trim(detail)
      throw .baseReferenceUnreadable(
        message: "git show \(baseReference):\(path) failed: "
          + (trimmed.isEmpty ? "unknown error" : trimmed),
      )
    }
  }

  /// The current JSONL text against its base-ref version.
  public static func checkAgainstBaseReference(
    baseReference: String,
    path: String,
    currentText: String,
    workingDirectory: String?,
    runner: some GitRunning = GitProcessRunner(),
  ) throws(GitError) -> AppendOnlyResult {
    let oldText = try readBaseReferenceFile(
      baseReference: baseReference,
      path: path,
      workingDirectory: workingDirectory,
      runner: runner,
    )
    return check(oldLines: splitJSONLines(oldText), newLines: splitJSONLines(currentText))
  }

  /// `/does not exist in|exists on disk, but not in/i`. Without the `u` flag a
  /// JavaScript `i` pattern folds only ASCII letters onto this ASCII pattern.
  private static func isAbsentAtBase(_ text: String) -> Bool {
    let folded = text.utf16.map { unit in
      (0x41 ... 0x5A).contains(unit) ? unit + 0x20 : unit
    }
    return ["does not exist in", "exists on disk, but not in"].contains { pattern in
      let needle = Array(pattern.utf16)
      guard folded.count >= needle.count else {
        return false
      }
      return (0 ... folded.count - needle.count).contains { start in
        CodeUnits.hasPrefix(folded, needle, at: start)
      }
    }
  }
}
